File: com/pdfjet/Bookmark.java
/*
 * Bookmark.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

import java.util.*;

/**
 * Please see Example_51 and Example_52
 */
public class Bookmark {
    /** The page this bookmark points to. */
    protected Page page = null;
    /** The y coordinate of the bookmark destination on the page. */
    protected float y = 0f;
    /** The object number of this bookmark. */
    protected int objNumber = 0;
    /** The number prefix added to the title by autoNumber. */
    protected String prefix = null;

    private int destNumber = 0;
    private String key = null;
    private String title = null;
    private Bookmark parent = null;
    private Bookmark prev = null;
    private Bookmark next = null;
    private List<Bookmark> children = null;
    private Destination dest = null;

    /**
     * Creates a bookmark.
     *
     * @param pdf the PDF.
     */
    public Bookmark(PDF pdf) {
        pdf.toc = this;
    }

    /**
     * Creates a bookmark and sets the location, key and title.
     *
     * @param page the page.
     * @param y the vertical location on the page.
     * @param key the key.
     * @param title the title.
     */
    private Bookmark(Page page, float y, String key, String title) {
        this.page = page;
        this.y = y;
        this.key = key;
        this.title = title;
    }

    /**
     * Add bookmark with the specified title to the page.
     *
     * @param page the page.
     * @param title the title.
     * @return the bookmark.
     */
    public Bookmark addBookmark(Page page, Title title) {
        Bookmark bm = this;
        while (bm.parent != null) {
            bm = bm.getParent();
        }
        String key = bm.next();
        Bookmark bookmark2 = new Bookmark(
                page,
                title.textLine.getDestinationY(),
                key,
                title.textLine.text.replaceAll("\\s+", " "));
        bookmark2.parent = this;
        bookmark2.dest = page.addDestination(key, title.textLine.getDestinationY());
        if (children == null) {
            children = new ArrayList<Bookmark>();
        } else {
            bookmark2.prev = children.get(children.size() - 1);
            children.get(children.size() - 1).next = bookmark2;
        }
        children.add(bookmark2);
        return bookmark2;
    }

    /**
     * Returns the destination key.
     *
     * @return the destination key.
     */
    public String getDestKey() {
        return this.key;
    }

    /**
     * Returns the bookmark title.
     *
     * @return the bookmark title.
     */
    public String getTitle() {
        return this.title;
    }

    /**
     * Returns the bookmark parent.
     *
     * @return the bookmark parent.
     */
    public Bookmark getParent() {
        return this.parent;
    }

    /**
     * Auto number the bookmark.
     *
     * @param textLine the text line.
     * @return the bookmark.
     */
    public Bookmark autoNumber(TextLine textLine) {
        Bookmark bm = getPrevBookmark();
        if (bm == null) {
            bm = getParent();
            if (bm.prefix == null) {
                prefix = "1";
            } else {
                prefix = bm.prefix + ".1";
            }
        } else {
            if (bm.prefix == null) {
                if (bm.getParent().prefix == null) {
                    prefix = "1";
                } else {
                    prefix = bm.getParent().prefix + ".1";
                }
            } else {
                int index = bm.prefix.lastIndexOf('.');
                if (index == -1) {
                    prefix = String.valueOf(Integer.parseInt(bm.prefix) + 1);
                } else {
                    prefix = bm.prefix.substring(0, index) + ".";
                    prefix += String.valueOf(Integer.parseInt(bm.prefix.substring(index + 1)) + 1);
                }
            }
        }
        textLine.setText(prefix);
        title = prefix + " " + title;
        return this;
    }

    /**
     * Returns this bookmark and all of its descendants in breadth-first order.
     *
     * @return the list of bookmarks.
     */
    protected List<Bookmark> toArrayList() {
        List<Bookmark> list = new ArrayList<Bookmark>();
        Queue<Bookmark> queue = new java.util.LinkedList<Bookmark>();
        int objNumber = 0;
        queue.add(this);
        while (!queue.isEmpty()) {
            Bookmark bm = queue.poll();
            bm.objNumber = objNumber++;
            list.add(bm);
            if (bm.getChildren() != null) {
                queue.addAll(bm.getChildren());
            }
        }
        return list;
    }

    /**
     * Returns the child bookmarks.
     *
     * @return the children, or null if there are none.
     */
    protected List<Bookmark> getChildren() {
        return this.children;
    }

    /**
     * Returns the previous sibling bookmark.
     *
     * @return the previous bookmark.
     */
    protected Bookmark getPrevBookmark() {
        return this.prev;
    }

    /**
     * Returns the next sibling bookmark.
     *
     * @return the next bookmark.
     */
    protected Bookmark getNextBookmark() {
        return this.next;
    }

    /**
     * Returns the first child bookmark.
     *
     * @return the first child.
     */
    protected Bookmark getFirstChild() {
        return this.children.get(0);
    }

    /**
     * Returns the last child bookmark.
     *
     * @return the last child.
     */
    protected Bookmark getLastChild() {
        return children.get(children.size() - 1);
    }

    /**
     * Returns the destination of this bookmark.
     *
     * @return the destination.
     */
    protected Destination getDestination() {
        return this.dest;
    }

    private String next() {
        destNumber++;
        return "dest#" + destNumber;
    }
}   // End of Bookmark.java

File: net/pdfjet/Bookmark.cs
/*
 * Bookmark.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Text.RegularExpressions;
using System.Collections.Generic;

namespace PDFjet.NET {
/// <summary>
/// Please see Example_48
/// </summary>
public class Bookmark {
    private int destNumber = 0;
    private Page page = null;
    private float y = 0f;
    private String key = null;
    private String title = null;
    private Bookmark parent = null;
    private Bookmark prev = null;
    private Bookmark next = null;
    private List<Bookmark> children = null;
    private Destination dest = null;
    internal int objNumber = 0;
    internal String prefix = null;

    /// <summary>Creates the root bookmark of the document outline.</summary>
    public Bookmark(PDF pdf) {
        pdf.toc = this;
    }

    private Bookmark(Page page, float y, String key, String title) {
        this.page = page;
        this.y = y;
        this.key = key;
        this.title = title;
    }

    /// <summary>
    /// Adds a bookmark with the specified title that points to the page.
    /// </summary>
    /// <param name="page">the page.</param>
    /// <param name="title">the title.</param>
    /// <returns>the new bookmark.</returns>
    public Bookmark AddBookmark(Page page, Title title) {
        Bookmark bm = this;
        while (bm.parent != null) {
            bm = bm.GetParent();
        }
        String key = bm.Next();

        Bookmark bookmark = new Bookmark(
                page, title.textLine.GetDestinationY(), key, Regex.Replace(title.textLine.text, @"\s+"," "));
        bookmark.parent = this;
        bookmark.dest = page.AddDestination(key, title.textLine.GetDestinationY());
        if (children == null) {
            children = new List<Bookmark>();
        } else {
            bookmark.prev = children[children.Count - 1];
            children[children.Count - 1].next = bookmark;
        }
        children.Add(bookmark);
        return bookmark;
    }

    /// <summary>Returns the destination key of this bookmark.</summary>
    public String GetDestKey() {
        return this.key;
    }

    /// <summary>Returns the title of this bookmark.</summary>
    public String GetTitle() {
        return this.title;
    }

    /// <summary>Returns the parent bookmark.</summary>
    public Bookmark GetParent() {
        return this.parent;
    }

    /// <summary>Numbers this bookmark by its position, for example 1.2, and adds the number to the title.</summary>
    public Bookmark AutoNumber(TextLine text) {
        Bookmark bm = GetPrevBookmark();
        if (bm == null) {
            bm = GetParent();
            if (bm.prefix == null) {
                prefix = "1";
            } else {
                prefix = bm.prefix + ".1";
            }
        } else {
            if (bm.prefix == null) {
                if (bm.GetParent().prefix == null) {
                    prefix = "1";
                } else {
                    prefix = bm.GetParent().prefix + ".1";
                }
            } else {
                int index = bm.prefix.LastIndexOf('.');
                if (index == -1) {
                    prefix = (Int32.Parse(bm.prefix) + 1).ToString();
                } else {
                    prefix = bm.prefix.Substring(0, index) + ".";
                    prefix += (Int32.Parse(bm.prefix.Substring(index + 1)) + 1).ToString();
                }
            }
        }
        text.SetText(prefix);
        title = prefix + " " + title;
        return this;
    }

    internal List<Bookmark> ToArrayList() {
        List<Bookmark> list = new List<Bookmark>();
        List<Bookmark> queue = new List<Bookmark>();
        queue.Add(this);
        int objNumber = 0;
        while (queue.Count > 0) {
            Bookmark bm = queue[0];
            queue.RemoveAt(0);
            bm.objNumber = objNumber++;
            list.Add(bm);
            List<Bookmark> children = bm.GetChildren();
            if (children != null) {
                foreach (Bookmark bm2 in children) {
                    queue.Add(bm2);
                }
            }
        }
        return list;
    }

    internal List<Bookmark> GetChildren() {
        return this.children;
    }

    internal Bookmark GetPrevBookmark() {
        return this.prev;
    }

    internal Bookmark GetNextBookmark() {
        return this.next;
    }

    internal Bookmark GetFirstChild() {
        return this.children[0];
    }

    internal Bookmark GetLastChild() {
        return children[children.Count - 1];
    }

    internal Destination GetDestination() {
        return this.dest;
    }

    private String Next() {
        ++destNumber;
        return "dest#" + destNumber.ToString();
    }
}   // End of Bookmark.cs
}   // End of namespace PDFjet.NET
